### PR TITLE
Issue #7176: Add AnnotatedDeclarationVisibilityCheck

### PR DESCRIPTION
**Pull request by gianmarcoschifone**
Issue #7176

Contribution repo PR: https://github.com/checkstyle/contribution/pull/1042

New module config: https://gist.githubusercontent.com/gianmarcoschifone/45d9200df8159aec19512b81c377eb04/raw/404ef8338df7fa7519cd21e5af8f7d5741d848b9/annotateddeclarationvisibility.xml

### CODE REVIEW

**Comment by gianmarcoschifone** [comment]
GitHub, generate website

**Comment by gianmarcoschifone** [comment]
GitHub, generate report

**Comment by gianmarcoschifone** [comment]
GitHub, generate website

**Comment by gianmarcoschifone** [comment]
Hi @romani  , PR is ready for review.
I analyzed the diff report and violations are regular.

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/7170a03_2026172350/reports/diff/index.html

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/f5dc96b_20260310193109/checks/modifier/annotatedmethodvisibilitymodifier.html#

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedMethodVisibilityModifierCheck.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+ * have only allowed visibility modifiers.
+ * </div>
+ *
+ * <p>
+ * This check enforces consistency between annotation presence and
+ * declared visibility. If a configured annotation is found on a target
+ * element, its visibility modifier must match one of the allowed values.
+ * </p>
+ *
+ * @since 13.4.0
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedMethodVisibilityModifierCheck.java]: 13.5.0

**Review by romani — CHANGES_REQUESTED** [review]
please rebase.

items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {
+            // annotation name resolution
+            TokenTypes.PACKAGE_DEF,
+            TokenTypes.IMPORT,
+            // tokens that can have annotations
+            TokenTypes.CLASS_DEF,
+            TokenTypes.INTERFACE_DEF,
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java:146]: something wrong with name of Check.

Method Visibility  vs target Class.

**Review by romani — CHANGES_REQUESTED** [review]
if we fully targeting to Methods, do we need extra tokens for types ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java (hunk 2 above)
Comment (gianmarcoschifone) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java:146]: The check should target any token that can have a visibility modifier, not just method, as reported in https://github.com/checkstyle/checkstyle/issues/7176.
Could we rename the check to `AnnotatedVisibilityModifier`?

**Comment by romani** [comment]
@gianmarcoschifone , please rebase and make it versioned 13.8.0 , I will make sue review if finished in July

**Comment by gianmarcoschifone** [comment]
@romani done

**Comment by romani** [comment]
CI is not happy , please address

**Comment by gianmarcoschifone** [comment]
@romani I think this check implementation will need some changes because what I tried was attempting to handle unqualified annotations through package and imports, this brought me to add non compilable resources. What do you think?

**Comment by mahfouz72** [comment]
@gianmarcoschifone Let's get the energy back into this PR. First, we need to rethink the name, since check can target everything, not just methods. Can its name be `AnnotatedDeclarationVisibilityCheck`. Please rename and rebase and we can start talking about implementation

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    private final Set<String> annotations = new HashSet<>(
+            Set.of("com.google.common.annotations.VisibleForTesting"));
+
+    /**
+     * Allowed visibility values.
+     * Acceptable values: public, protected, package-private, private.
+     */
+    private final Set<String> visibility = new HashSet<>(
+            Set.of(PROTECTED, PACKAGE_PRIVATE));
+
```
Comment (mahfouz72) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java]: https://checkstyle.sourceforge.io/property_types.html#AccessModifierOption.5B.5D

Can we use this property type? Please see how other checks use it

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+AnnotatedDeclarationVisibility
+annotations = org.jspecify.annotations.Nullable, org.junit.jupiter.api.Disabled
+visibility = (default)protected,package-private
+tokens = (default)PACKAGE_DEF, IMPORT, CLASS_DEF, \
+         INTERFACE_DEF, ENUM_DEF, RECORD_DEF, \
+         METHOD_DEF, CTOR_DEF, VARIABLE_DEF, ANNOTATION_DEF,
+
+
+*/
+// non-compiled because it's not a valid annotation reference, since the import is not defined
```
Comment (mahfouz72) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: I don't get this, why can't we import valid annotations and make the file compiled?

**Review by mahfouz72 — COMMENTED** [review]
items

**Comment by mahfouz72** [comment]
@gianmarcoschifone Can you make CI green and generate report

**Comment by mahfouz72** [comment]
GitHub, generate website

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/*xml
+<module name="Checker">
+  <module name="TreeWalker">
+    <module name="AnnotatedDeclarationVisibility">
+      <property name="tokens" value="PACKAGE_DEF, IMPORT, METHOD_DEF"/>
```
Comment (mahfouz72) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]: We shouldn't add `PACKAGE_DEF` and `IMPORT` here, right? These are required tokens that the check will always stop on

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java (hunk 4 above)
Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: because it is used to verify this:

```java
@Override
public void beginTree(DetailAST rootAST) {
    importedAnnotations.clear();
    starImports.clear();
}
```

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java (hunk 5 above)
Comment (gianmarcoschifone) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]: Yes those tokens should always be present but that example was showing a custom tokens config for using only `METHOD_DEF`, if I just don't add them will it work because they are in required tokens?

Comment (mahfouz72) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]: Can you please try it I don't remember

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheckTest.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    @Test
+    public void testCustomVisibility() throws Exception {
+        final String[] expected = {
+            "21:5: " + getCheckMessage(MSG_KEY, "protected"),
+            "25:5: " + getCheckMessage(MSG_KEY, "private"),
+            "29:10: " + getCheckMessage(MSG_KEY, "package-private"),
+            "36:5: " + getCheckMessage(MSG_KEY, "protected"),
+        };
```
Comment (gianmarcoschifone) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheckTest.java:116]: @mahfouz72 In my implementation I was logging the violation on the modifier when present while on the declaration ast when not present (package-private case).
Do you think this is ok or it is better to always log on the declaration ast to not have differences?

Comment (mahfouz72) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheckTest.java:116]: I prefer always logging on the declaration AST for consistency, especially since package-private has no modifier AST. Also the rule is about the declaration’s visibility as a whole, it is ok to use the declaration AST in all cases.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java (hunk 4 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: I am struggling to understand why we need to make the code non-compilable to kill the mutation if no execution with compilable files kills it?

Hint: We can execute multiple files in the same test run (You can find examples in codebase), so the mutation could potentially be killed through a combination of compilable files. 

Am I missing something?

Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: Yes, I am already testing this by executing two files in the same test run.
`InputAnnotatedDeclarationVisibilityClearedImport1` imports some annotations, while `InputAnnotatedDeclarationVisibilityClearedImport2` uses those annotations without importing them.
The purpose is to verify that the list of previously stored imports is cleared between files. If it is not cleared, the annotation in the second file would be incorrectly resolved using the imports collected from the first one.
For this reason, the second file is intentionally non-compilable on its own, since it references an annotation that it does not import.

Maybe there is a better way as you say, I will think about this

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java (hunk 5 above)
Comment (gianmarcoschifone) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]: yes it works, done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java (hunk 3 above)
Comment (gianmarcoschifone) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+     */
+    public void setVisibility(AccessModifierOption... values) {
+        visibility = Arrays.copyOf(values, values.length);
+    }
+
+    @Override
+    public void beginTree(DetailAST rootAST) {
+        importedAnnotations.clear();
+        starImports.clear();
+    }
```
Comment (mahfouz72) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java:122]: should we clean `currentPackage` here also to avoid wrong resolution in next file with the previous file package name?
Can we think of a test to kill this when pitest mutates it?

Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java:122]: For example, if `First.java `contains `package com.example; `and the next processed file has no package declaration, currentPackage will still be `com.example`. If the check is configured with `com.example.MyAnnotation`, then `@MyAnnotation` in the second file will be incorrectly resolved as `com.example.MyAnnotation` and may trigger a violation.

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java (hunk 4 above)
Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: done!

**Comment by gianmarcoschifone** [comment]
I'm not sure on how to proceed for fixing https://circleci.com/gh/checkstyle/checkstyle/1660427, maybe we should update `validation.sh`?

File: pom.xml [hunk 8]
```diff
@@ -4329,6 +4329,9 @@
               <targetTests>
                 <param>com.puppycrawl.tools.checkstyle.checks.modifier.*</param>
               </targetTests>
+              <avoidCallsTo>
+                <avoidCallsTo>com.puppycrawl.tools.checkstyle.utils.NullUtil</avoidCallsTo>
```
Comment (mahfouz72) [inline pom.xml:4335]: Why do we need this, can't it be killed by tests? Is this for cheker framework null check right?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java
Comment (mahfouz72) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java:1]: Is there example where this exact file mark annotations as vioaltion? If yes please point to it

I mean where we configure annotations property to flag the package defined annotations like:

`annotations = com.puppycrawl.tools.checkstyle.checks.modifier.annotateddeclarationvisibility.Disabled`

**Review by mahfouz72 — CHANGES_REQUESTED** [review]
Great Job! Direction is good.

Some comments:

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java (hunk 4 above)
Comment (mahfouz72) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClearedImport2.java]: Great!!

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java
Comment (mahfouz72) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java:1]: https://github.com/gianmarcoschifone/checkstyle/blob/53e6c1fee37a13ff9f13d22634c28e9c3b873d6f/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilitySamePackage.java

I found it

File: pom.xml (hunk 8 above)
Comment (gianmarcoschifone) [reply pom.xml:4335]: yes that is for checker framework. No it can't be killed by tests, I'm pretty sure this is the way, I used it often and saw it often

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java
Comment (gianmarcoschifone) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java:1]: all good?

Comment (mahfouz72) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityClear2.java:1]: Good!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java (hunk 7 above)
Comment (gianmarcoschifone) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/AnnotatedDeclarationVisibilityCheck.java:122]: done, a suppression in `checkstyle-resources-suppressions.xml‎` is needed for not having a package declaration. I saw other checks doing it also.

**Comment by mahfouz72** [comment]
Github, generate site

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+tokens = (default)PACKAGE_DEF, IMPORT, CLASS_DEF, \
+         INTERFACE_DEF, ENUM_DEF, RECORD_DEF, \
+         METHOD_DEF, CTOR_DEF, VARIABLE_DEF, ANNOTATION_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+import com.google.common.annotations.VisibleForTesting;
```
Comment (mahfouz72) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: Files tagged `Compilable with Java25` are compiled by `.ci/validation.sh` with a bare j`avac --release 25 -d target <file>` (no -classpath), so only the JDK is     
  available. Guava isn't on that classpath, so the  import can't resolve.


  Alternative solution for now: drop the Guava import and declare the annotation locally in the file or use any other available jdk import

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  @VisibleForTesting
+  protected void protectedMethod() {}
+
+  // ok, default visibility includes package-private
+  @VisibleForTesting
+  void packagePrivateMethod() {}
+
+  // ok, default visibility includes protected
+  @VisibleForTesting
+  protected int protectedField;
```
Comment (mahfouz72) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]:
```suggestion
  // ok, fields are not checked when tokens only include METHOD_DEF  
  @VisibleForTesting
  protected int protectedField;
```

**Review by mahfouz72 — CHANGES_REQUESTED** [review]
Please make CI green and generate report

**Comment by mahfouz72** [comment]
> Module AnnotatedDeclarationVisibility not found in contribution PR patch.
Please add AnnotatedDeclarationVisibility to only one of the following files:
   checkstyle-tester/checks-nonjavadoc-error.xml
or checkstyle-tester/checks-only-javadoc-error.xml

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java (hunk 9 above)
Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: problem is that if I change the config of the check then CI complains that there is no example with default config values

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java (hunk 10 above)
Comment (gianmarcoschifone) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example2.java]: done, sorry

**Comment by gianmarcoschifone** [comment]
GitHub, generate report

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java (hunk 9 above)
Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: we can update junit validation to work more smartly.
Validation is in initial phase now only, feel free to improve.

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: @gianmarcoschifone , can we use other annoations to avoid classpath dependency ?

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: can we add two Inputs ? default and nondefault ?

Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: @romani there must be one Compact Source File with check's default configuration, default is the Guava one though. There is already also the non default one. Do you mean that in the default one I just don't put any usage of the annotation?

Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: yes, jsut make Input to make validation happy.
you have use any other annotation. 
Default is config, but Input and behavior might not need to have violation, just some java file.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java [hunk 11]
```diff
@@ -163,6 +163,7 @@ public class XdocsExamplesAstConsistencyTest {
         "checks/javadoc/writetag",
         "checks/metrics/classdataabstractioncoupling",
         "checks/metrics/cyclomaticcomplexity",
+        "checks/modifier/annotateddeclarationvisibility",
```
Comment (mahfouz72) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java]: is this necessary? can we make examples identical as required by this tests?

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java]: no more examples degradation.... all PRs must comply.

Comment (gianmarcoschifone) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsExamplesAstConsistencyTest.java]: done

**Comment by gianmarcoschifone** [comment]
Github, generate website

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java (hunk 9 above)
Comment (gianmarcoschifone) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/compact/InputAnnotatedDeclarationVisibilityCompactSourceFile.java]: done

File: src/site/xdoc/checks/modifier/annotateddeclarationvisibility.xml.template [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        </p>
+        <macro name="example">
+          <param name="path"
+                 value="resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example3.java"/>
+          <param name="type" value="config"/>
+        </macro>
+        <p id="Example3-code">Example:</p>
+        <macro name="example">
+          <param name="path"
+                 value="resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/Example3.java"/>
```
Comment (mahfouz72) [inline src/site/xdoc/checks/modifier/annotateddeclarationvisibility.xml.template:72]: In the Examples Java files, we have up to example 4, but we didn't add it here

**Review by mahfouz72 — CHANGES_REQUESTED** [review]
Everything looks good, last for me

File: src/site/xdoc/checks/modifier/annotateddeclarationvisibility.xml.template (hunk 12 above)
Comment (gianmarcoschifone) [reply src/site/xdoc/checks/modifier/annotateddeclarationvisibility.xml.template:72]: done

**Comment by gianmarcoschifone** [comment]
Github, generate website

**Review by mahfouz72 — APPROVED** [review]
Ok to merge, thanks a lot!!

File: src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+        final String[] expectedViolation = {
+            "5:5: " + getCheckMessage(AnnotatedDeclarationVisibilityCheck.class,
+                    AnnotatedDeclarationVisibilityCheck.MSG_KEY, "public"),
+        };
+
+        final String methodXpath =
+            "/COMPILATION_UNIT/CLASS_DEF"
+                + "[./IDENT[@text='InputXpathAnnotatedDeclarationVisibilityPublic']]"
+                + "/OBJBLOCK/METHOD_DEF[./IDENT[@text='violationMethod']]";
```
Comment (romani) [inline src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java:61]: goal of 3 test of Xpath to show diversity of Xpath.
now we have 3 "equal" xpaths, lets put violation in inner class or annon object to make Xpath be different.
If there is any other way to make it different I am ok.

**Review by romani — CHANGES_REQUESTED** [review]
DONE: ~please point me to~ Input where user target annotation is default - VisibleForTessting
but in code import if for other annotation of the same name, in code user use short form of annotation. In such case there should be no violation, as Check should recognize that it is same by short form by not a target.

TODO:

> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/e4c19f4_2026142508/reports/diff/index.html

very interesting results, Please sen PRs to Comunda and Hbase to udpate some (not all), to see how users will react on such updates. If both PRs merged with pleasure, it means we doing right thing with such Check. If they do not like it, it is good feeback on Checks design.

itms:

File: src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java (hunk 13 above)
Comment (mahfouz72) [reply src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java:61]: @romani 
> please point me to Input where user target annotation is default - VisibleForTessting
but in code import if for other annotation of the same name, in code user use short form of annotation. In such case there should be no violation,

https://github.com/gianmarcoschifone/checkstyle/blob/1621291d85ac8e0e13d0703db378036c696cb87a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/modifier/annotateddeclarationvisibility/InputAnnotatedDeclarationVisibilityMultipleConfigured.java

Comment (mahfouz72) [reply src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java:61]:
```Java
/*
AnnotatedDeclarationVisibility
annotations = com.google.common.annotations.VisibleForTesting,java.lang.Deprecated
visibility = (default)protected,package
tokens = (default)PACKAGE_DEF, IMPORT, CLASS_DEF, \
         INTERFACE_DEF, ENUM_DEF, RECORD_DEF, \
         METHOD_DEF, CTOR_DEF, VARIABLE_DEF, ANNOTATION_DEF,

*/

package com.puppycrawl.tools.checkstyle.checks.modifier.annotateddeclarationvisibility;

import org.fest.util.VisibleForTesting;

public class InputAnnotatedDeclarationVisibilityMultipleConfigured {

    ....
    
    // violation below 'Annotated element has disallowed visibility 'private'.'
    @com.google.common.annotations.VisibleForTesting
    private int violationPrivateField;

    @com.google.common.annotations.VisibleForTesting
    protected InputAnnotatedDeclarationVisibilityMultipleConfigured() {}

    @VisibleForTesting
    public InputAnnotatedDeclarationVisibilityMultipleConfigured(int x) {}

  

    @VisibleForTesting
    protected int allowedField2;
}
```

Comment (romani) [reply src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java:61]: cleared on Input question, but this thread is on Xpath, so should be done.

Comment (gianmarcoschifone) [reply src/it/java/org/checkstyle/suppressionxpathfilter/modifier/XpathRegressionAnnotatedDeclarationVisibilityTest.java:61]: done the xpath update

**Comment by romani** [comment]
ready for merge
ok, only feedback from users is required before we merge.

**Comment by gianmarcoschifone** [comment]
@romani sent to camunda. Hbase current master doesn't have any violations. We can wait for camunda's response

**Comment by romani** [comment]
If no response on Monday, July  20, we will merge.

**Review by romani — APPROVED** [review]
thanks a lot !!
